kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -355,236 +354,197 @@ uint64_t KvbAppFilter::newestTagSpecificPublicEventGroupId() const {
-
-  // reset the vectors to read the new batch from storage
-  eg_state->event_group_id_batch.clear();
-  eg_state->it = eg_state->event_group_id_batch.end();
-  // holds a maximum of kBatchSize ordered public event group ids at a time
-  std::vector<uint64_t> public_event_group_ids{};
-  // holds a maximum of kBatchSize ordered private event group ids at a time
-  std::vector<uint64_t> private_event_group_ids{};
-
-  // populate public_event_group_ids
-  if (eg_state->public_offset != 0) {
-    // start reading from the offset until a complete batch is read or no more event groups exist for the tag
-    for (uint64_t i = eg_state->public_offset; i < std::min(eg_state->public_offset + kBatchSize, public_end + 1);
-         ++i) {
-      // get global_event_group_id and external_tag_event_group_id corresponding to tag_event_group_id
-      // tag + kTagTableKeySeparator + latest_tag_event_group_id concatenation is used as key for kv-updates of type
-      // kExecutionEventGroupTagCategory
-      const auto &[global_eg_id, _] =
-          getValueFromTagTable(kPublicEgId + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(i));
-      (void)_;
-      public_event_group_ids.emplace_back(global_eg_id);
-    }
-  }
-  // populate private_event_group_ids
-  if (eg_state->private_offset != 0) {
-    for (uint64_t i = eg_state->private_offset; i < std::min(eg_state->private_offset + kBatchSize, private_end + 1);
-         ++i) {
-      const auto &[global_eg_id, _] =
-          getValueFromTagTable(client_id_ + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(i));
-      (void)_;
-      private_event_group_ids.emplace_back(global_eg_id);
+
+  if (not private_start) {
+    // requested external event group id == public event group id
+    uint64_t global_id;
+    std::tie(global_id, std::ignore) = getValueFromTagTable(
+        kPublicEgId + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(external_event_group_id));
+    return {global_id, false, private_end ? private_end - 1 : 0, external_event_group_id};
+  } else if (not public_start) {
+    // requested external event group id == private event group id
+    uint64_t global_id;
+    std::tie(global_id, std::ignore) = getValueFromTagTable(
+        client_id_ + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(external_event_group_id));
+    return {global_id, true, external_event_group_id, public_end ? public_end - 1 : 0};
+  }
+
+  std::string prefix = client_id_ + kTagTableKeySeparator;
+
+  // Binary search in private event groups
+  uint64_t window_size;
+  auto window_start = private_start;
+  auto window_end = private_end;
+
+  // Cursors inside the search window; All point to the same entry
+  // client_id_ # current_pvt_eg_id => current_global_eg_id # current_ext_eg_id
+  auto [current_pvt_eg_id, current_global_eg_id, current_ext_eg_id] = std::make_tuple(0ull, 0ull, 0ull);
```

Comment:
I was wondering why we need to create a tuple and then bind the variables to it. Can we just create 3 separate variables instead?